core/src/cache.rs

```diff
@@ -1746,6 +2107,525 @@ pub mod resolvers {
+            std::mem::transmute::<
+                &'borrow mut HashMap<FileId, &'static Ast<'static>>,
+                &'borrow mut HashMap<FileId, &'ast Ast<'ast>>,
+            >(asts)
+        }
+
+        /// Same as [Self::borrow_ast], but for the typing context. Since contexts are cheaply
+        /// cloneable and needs to be provided as owned variant, [Self::borrow_clone_context]
+        /// returns a clone.
+        ///
+        /// # Safety
+        ///
+        /// The caller must ensure that the asts stored in `self.asts` have been allocated with
+        /// `_alloc` (or any allocator that will live as long as `_alloc`).
+        unsafe fn borrow_context_copy<'ast>(
+            _alloc: &'ast AstAlloc,
+            type_ctxt: &typecheck::Context<'static>,
+        ) -> typecheck::Context<'ast> {
+            std::mem::transmute::<typecheck::Context<'static>, typecheck::Context<'ast>>(
+                type_ctxt.clone(),
+            )
+        }
+
+        /// Same as [Self::borrow_clone_context] but return a mutable reference.
+        ///
+        /// # Safety
+        ///
+        /// 1. The caller must ensure that the asts stored in `self.asts` have been allocated with
+        ///    `_alloc` (or any allocator that will live as long as `_alloc`).
+        /// 2. The caller must ensure that any new type or AST component inserted must've been
+        ///    allocated with `_alloc`, or at least be guaranteed to live as long as `_alloc`.
+        unsafe fn borrow_context_mut<'ast, 'borrow>(
+            _alloc: &'ast AstAlloc,
+            type_ctxt: &'borrow mut typecheck::Context<'static>,
+        ) -> &'borrow mut typecheck::Context<'ast>
+        where
+            'ast: 'borrow,
+        {
+            std::mem::transmute::<
+                &'borrow mut typecheck::Context<'static>,
+                &'borrow mut typecheck::Context<'ast>,
+            >(type_ctxt)
+        }
+
+        /// Parses a Nickel expression and stores the corresponding AST in the cache.
+        pub fn parse_nickel<'ast>(
+            &'ast mut self,
+            file_id: FileId,
+            source: &str,
+        ) -> Result<&'ast Ast<'ast>, ParseErrors> {
+            let ast = parse_nickel(&self.alloc, file_id, source)?;
+            let ast = self.alloc.alloc(ast);
+            // Safety: we are transmuting the lifetime of the AST from `'ast` to `'static`. This is
+            // unsafe in general, but we never use or leak any `'static` reference. It's just a
+            // placeholder. We only store such `Ast<'static>` in `asts`, and return them as `'a`
+            // references where `self: 'a` in `get()` or the various `borrow_xxx` methods.
+            //
+            // Thus, the `'static` lifetime isn't observable from outsideof `AstCache`, and asts
+            // are never used as `Ast<'static>` from within the implementation of `AstCache`. They
+            // are always tied to the lifetime of `self.alloc` through `Self::get2`.
```

Comment:
get2?

Reply:
I wasn't very inspired during the first pass :sweat_smile: it's one of `borrow_xxx` now. Will update.